### PR TITLE
Fixes issues with streaming + sequence ID

### PR DESCRIPTION
**Pull request by elijahbenizzy**
https://github.com/DAGWorks-Inc/burr/issues/177

This ensures the following:

1. That increment_sequence_id is always called first
2. That it is only called once
3. That all pre run step hooks are called only once
4. That all post run step hooks are called only once

We'll likely want to improve this, the control flow is complex.

- [x] PR has an informative and human-readable title (this will be pulled into the release notes)
- [x] Changes are limited to a single goal (no scope creep)
- [x] Code passed the pre-commit check & code is left cleaner/nicer than when first encountered.
- [x] Any _change_ in functionality is tested
- [x] New functions are documented (with a description, list of inputs, and expected output)
- [x] Placeholder code is flagged / future TODOs are captured in comments
- [x] Project documentation has been updated if adding/changing functionality.

### CODE REVIEW

**Comment by elijahbenizzy** [comment]
Decided to add to existing testss, as these are more "lifecycle tests" than unit tests. The logic is complicated and works on state, so there's not that much more we can do to break it up...

File: burr/core/application.py [hunk 1]
```diff
@@ -867,24 +874,28 @@ def callback(
                 result: Optional[dict],
                 state: State,
                 exc: Optional[Exception] = None,
-                seq_id=self.sequence_id,
+                # seq_id=self.sequence_id,
```
Comment (skrawcz) [inline burr/core/application.py:877]: this supposed to be commented?

Comment (skrawcz) [reply burr/core/application.py:877]: delete otherwise?

File: burr/core/application.py [hunk 2]
```diff
@@ -836,18 +848,10 @@ def streaming_response(state: State, prompt: str) -> Generator[dict, None, Tuple
             # For context, this is specifically for the case in which you want to have
             # multiple terminal points with a unified API, where some are streaming, and some are not.
             if next_action.name in halt_before and next_action.name not in halt_after:
-                self._adapter_set.call_all_lifecycle_hooks_sync(
-                    "post_run_step",
```
Comment (skrawcz) [inline burr/core/application.py:840]: I assume this is being called somewhere else, and this was being called twice?

**Review by skrawcz — APPROVED** [review]
I assume it works!

File: burr/core/application.py (hunk 1 above)
Comment (elijahbenizzy) [reply burr/core/application.py:877]: Yep, not meant to be there (it made things more complicated)

File: burr/core/application.py (hunk 2 above)
Comment (elijahbenizzy) [reply burr/core/application.py:840]: Yep, we delegate to `run`

**Pull request merged**